### PR TITLE
Use nested coproduct list support from Iota 0.2.x

### PR DESCRIPTION
**Pull request by andyscott**
~This build will fail because the docs depend on the integration library which depends on the core macros for generating the free boilerplate. I've added a type to the `@free` generated code, so the integration library build is now outdated.~

### CODE REVIEW

File: freestyle/shared/src/main/scala/freestyle/module.scala [hunk 1]
```diff
@@ -100,66 +86,6 @@ final class moduleImpl(val c: Context) {
     cls.impl.body.collect {
       case valDef: ValDef if valDef.rhs.isEmpty => (valDef.name, valDef.tpt) }
 
-  private[this] def expandEffectType(tpe: Type): List[Type] = {
```
Comment (raulraja) [inline freestyle/shared/src/main/scala/freestyle/module.scala:103]: 😍

Comment (andyscott) [reply freestyle/shared/src/main/scala/freestyle/module.scala:103]: total lines changed `+22 −85` 😉

File: freestyle/shared/src/main/scala/freestyle/module.scala [hunk 2]
```diff
@@ -171,8 +97,24 @@ final class moduleImpl(val c: Context) {
       Template(parentTrees, self, body0)).validNel
   }
 
+  private[this] def makeCoproductTree(effectTuples: List[(Name, Tree)]): Tree = {
```
Comment (diesalbla) [inline freestyle/shared/src/main/scala/freestyle/module.scala]: * ¿Can we do the `map (_._2)` in the call to `makeCoproductTree`? 
* ¿Can we narrow down the type of the `Tree` to what `.tpt` is?

File: freestyle/shared/src/main/scala/freestyle/module.scala [hunk 3]
```diff
@@ -65,24 +65,10 @@ final class moduleImpl(val c: Context) {
       cls             <- decodeAnnotationTarget(annottees.toList)
       effectTuples     = gatherEffects(cls)
       classTree       <- makeClassTree(cls)
-      objectTree      <- makeModuleTree(cls, effectTuples)
+      coproductTree    = makeCoproductTree(effectTuples)
```
Comment (diesalbla) [inline freestyle/shared/src/main/scala/freestyle/module.scala]: You can move the calls to `gatherEffects` and `makeCoproductTree` inside `makeModuleTree`.

**Review by diesalbla — APPROVED** [review]
Great cleanup.

File: freestyle/shared/src/main/scala/freestyle/free.scala [hunk 4]
```diff
@@ -141,6 +141,7 @@ object freeImpl {
 
           import _root_.freestyle.{ FreeS, FSHandler, InjK }
 
+          type OpTypes = _root_.iota.KCons[Op, _root_.iota.KNil]
```
Comment (diesalbla) [inline freestyle/shared/src/main/scala/freestyle/free.scala]: Why do we need to change the `@free` macro? The `Op` is a constant name, not a fresh type name.

**Comment by andyscott** [comment]
Once we eventually have this PR passing, is there an easy way to compile the documentation against this release before we merge (to make sure it's all working 👍 )?

**Comment by juanpedromoreno** [comment]
@andyscott It's already being checked: https://github.com/frees-io/freestyle/blob/feature/iota-nested-coproducts/.travis.yml#L30

**Review by juanpedromoreno — APPROVED** [review]

**Comment by juanpedromoreno** [comment]
I'm merging since current issues are related to coursier cache. Checked locally and it's working as expected.

**Pull request merged**